### PR TITLE
Release tile keys from protected list.

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Release keys, which was marked before as protected. After releasing tile
keys associated data in cache could expire or be evicted.

For releasing tile keys we should release quad tree and tile data.
This means find quad tree which contains data hanle for our key,
release this quad tree only in case if no other protected tiles depend
on this quad tree. For each tile release data under data handle.

Added unit test with releasing 2 tiles keys under common quad tree,
which has 3 protected tile keys, so quad tree associated key should not
be released. Other test tries to release last tile key for quad tree,
so 2 keys should be released. Test which tries to relese not protected
key.

Relates-To: OLPEDGE-2127

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp [hunk 1]
```diff
@@ -751,12 +754,12 @@ std::string DefaultCacheImpl::GetExpiryKey(const std::string& key) const {
 bool DefaultCacheImpl::Protect(const DefaultCache::KeyListType& keys) {
   std::lock_guard<std::mutex> lock(cache_lock_);
   auto start = std::chrono::steady_clock::now();
-  protected_keys_.Protect(keys, [&](const std::string& key) {
+  auto result = protected_keys_.Protect(keys, [&](const std::string& key) {
     if (!RemoveKeyLru(key)) {
```
[the remark this answers is not in the document]
Comment (LiubovDidkivska) [reply olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp:758]: this is lambda, which is called in ProtectedKeyList::Protect

File: olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp [hunk 2]
```diff
@@ -654,6 +654,106 @@ bool VersionedLayerClientImpl::Protect(const TileKeys& tiles) {
...
+
+  cache::KeyValueCache::KeyListType keys_to_release;
+
+  std::map<geo::TileKey, std::map<geo::TileKey, std::string>>
+      quad_trees_with_protected_tiles;
+
+  auto process_tile = [&](const geo::TileKey& tile_key_to_release) {
+    auto max_depth =
+        std::min<std::uint32_t>(tile_key_to_release.Level(), kQuadTreeDepth);
+    for (int i = max_depth; i >= 0; --i) {
```
[the remark this answers is not in the document]
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp]: max_depth changed to signed, due to check >=0

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    : layer_id_(layer_id),
+      version_(version),
+      cache_(settings.cache),
+      data_cache_repository_(catalog, settings.cache),
+      partitions_cache_repository_(catalog, settings.cache),
+      repository_(catalog, settings),
+      quad_trees_with_protected_tiles_(),
+      keys_to_release_() {}
+
+cache::KeyValueCache::KeyListType&&
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp:57]: Why do you return a rvalue when you can simply return the cache::KeyValueCache::KeyListType which is the same except you are not forced to use std::move.
Also the name of the method will get much shorter: `GetKeysToRelease`

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+ReleaseDependencyResolver::GenerateAndMoveKeysToRelease(const TileKeys& tiles) {
+  for (const auto& tile : tiles) {
+    if (!ProcessTileKeyInMap(tile)) {
+      ProcessTileKeyInCache(tile);
+    }
+  }
+  return std::move(keys_to_release_);
+}
+
+bool ReleaseDependencyResolver::ProcessTileKeyInMap(
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp:67]:
```suggestion
bool ReleaseDependencyResolver::ProcessTileKey(
```

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        quad_trees_with_protected_tiles_.erase(it);
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+ReleaseDependencyResolver::MapOfQuads::iterator
+ReleaseDependencyResolver::FindQuadInMap(const geo::TileKey& tile_key) {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp:95]:
```suggestion
ReleaseDependencyResolver::FindQuad(const geo::TileKey& tile_key) {
```

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+namespace {
+constexpr auto kQuadTreeDepth = 4;
+}  // namespace
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+cache::KeyValueCache::KeyListType
+ReleaseDependencyResolver::GenerateKeysToRelease(
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.cpp:36]:
```suggestion
ReleaseDependencyResolver::GetKeysToRelease(
```

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+#include "repositories/DataCacheRepository.h"
+#include "repositories/DataRepository.h"
+#include "repositories/PartitionsCacheRepository.h"
+#include "repositories/PartitionsRepository.h"
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+class ReleaseDependencyResolver {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h:38]: Consider also using this for Protect().
Means the class name should be ProtectionDependencyResolver.

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+ public:
+  // group quad trees and protected tiles. Check if for each quad tree we want
+  // to release all protected keys associated, if true add this quad tree to
+  // released keys
+  static cache::KeyValueCache::KeyListType GenerateKeysToRelease(
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h:43]: This class setup is very strange. A static method that creates a temporary instance of itself to generate some data and copies internally not very cheap data. If it is only temporary class then do not copy but use references and remove the static from this method.

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      const client::HRN& catalog, const std::string& layer_id,
+      const int64_t& version, const client::OlpClientSettings& settings,
+      const TileKeys& tiles);
+
+ private:
+  using MapOfTileDataKeys = std::map<geo::TileKey, std::string>;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h:49]: Aliases are usually suffixed by "Type", e.g. `DataKeysType`
As there is no difference between partition data or tile data key, both use data handles.

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  using MapOfQuads = std::map<geo::TileKey, MapOfTileDataKeys>;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h:50]:
```suggestion
  using QuadsType = std::map<geo::TileKey, MapOfTileDataKeys>;
```

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h (hunk 8 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h]: Static method removed

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  QuadsType::iterator FindQuad(const geo::TileKey& tile_key);
+
+  TilesDataKeysType ProcessQuad(const read::QuadTreeIndex& cached_tree,
+                                const geo::TileKey& tile);
+
+  void ProcessTileKeyInCache(const geo::TileKey& tile);
+
+ private:
+  std::string layer_id_;
+  int64_t version_;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h:66]:
```suggestion
  const std::string& layer_id_;
  const int64_t version_;
```

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+#include "repositories/PartitionsCacheRepository.h"
+#include "repositories/PartitionsRepository.h"
+
+namespace olp {
+namespace dataservice {
+namespace read {
+
+// group quad trees and protected tiles. Check if for each quad tree we want
+// to release all protected keys associated, if true add this quad tree to
+// released keys
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h:40]: This is not a proper comment. Use `///` for simple Doxygen comments.

File: olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp [hunk 13]
```diff
@@ -420,14 +420,15 @@ PartitionResponse PartitionsRepository::GetTile(
 }
 PORTING_POP_WARNINGS()
 
-bool PartitionsRepository::FindQuadTree(
-    const std::string& layer, boost::optional<int64_t> version,
-    const olp::geo::TileKey& tile_key, read::QuadTreeIndex& tree) {
+bool PartitionsRepository::FindQuadTree(const std::string& layer,
+                                        boost::optional<int64_t> version,
+                                        const olp::geo::TileKey& tile_key,
+                                        read::QuadTreeIndex& tree) {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp:426]: Why is this again formatted differently? I don't see anything's been changed.

File: olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+/// group quad trees and protected tiles. Check if for each quad tree we want
+/// to release all protected keys associated, if true add this quad tree to
+/// released keys
+class ReleaseDependencyResolver {
+ public:
+  ReleaseDependencyResolver(const client::HRN& catalog,
+                            const std::string& layer_id, const int64_t& version,
+                            const client::OlpClientSettings& settings,
+                            const client::ApiLookupClient& lookup_client);
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ReleaseDependencyResolver.h:46]: It seems to me like over-engineered to take in ApiLookupClient to just create a PartitionsRepository to just have a helper method for accessing the cache. Please ask yourself if the FindQuadTree really belongs in PartitionsRepositoy and not in PartitionsCacheRepository. I would place it in PartitionsCacheRepository.

**Pull request merged**